### ISSUE TITLE
How to publish MetadataAuditEvent

### ISSUE THREAD

**Issue by priyanka1555** [linkedin/datahub#1766]
Hello,
I've produced the MetadataChangeEvent, and it is persisted to MYSQL, now how can i use the MetadataAuditEvent?

**Comment by camelliazhang2018** [comment]
Hi @priyanka1555  MetadataAuditEvent should be automatically emitted after the change is accepted and committed, otherwise a FailedMetadataChangeEvent will be emitted instead.

**Comment by priyanka1555** [comment]
Do I have to make changes in MetadataChangeEvent.avsc file, to test the MetadataAuditEvent?

**Comment by mars-lan** [comment]
@priyanka1555 No need the play with MetadataChangeEvent.avsc. Like @camelliazhang2018 mentioned, if your mce-consumer-job has successfully processed the event, a corresponding MAE should be fired by GMS. The MAE is subsequently processed by mae-consumer-job to update the search index and/or graph.

**Comment by mars-lan** [comment]
Please also feel free to utilize our [Slack channel](https://github.com/linkedin/datahub/blob/master/docs/slack.md) as a more efficient way to troubleshoot issues like this.

**Comment by jplaisted** [comment]
To follow up our current slack conversation, check out the [DataHub Architecture diagram](https://github.com/linkedin/datahub/blob/master/docs/architecture/architecture.md).

tl;dr of MCE flow:
- You emit an MCE to Kafka
- `mce-consumer-job` will handle the event.
- `mce-consumer-job` will take the MCE and call the `gms` to ingest it.
- The `gms` (aka `metadata-store`) will then write to MySQL, and if that succeeds, it will emit an MAE to kafka.
- The `mae-consumer-job` will handle the event.
- `mae-consumer-job` will update elasticsearch.
- You can now search for your data on DataHub.

tl;dr of searching on DataHub flow (assuming the above is done):
- You go to DataHub and search for something.
- This calls the `gms` to perform the search.
- `gms` delegates to elasticsearch.
- Results are returned from elasticsearch and bubbled back up to DataHub.


So, as you can see, for this to work you need the entire DataHub stack up and running (DataHub, `mce-consumer-job`, `mae-consumer-job`, `gms`, and dependencies like MySQL, kafka, elasticsearch, etc). The easiest way to do this is with our [Docker quickstart script](https://github.com/linkedin/datahub/blob/master/docs/quickstart.md) (though note this does not use your local code, if you are trying to develop. It pulls our latest builds from dockerhub. I'm working on making docker work better with development, see https://github.com/linkedin/datahub/issues/1760).

---

Another anecdote: You don't want to be emitting MAEs yourself. You (or your metadata source applications / crawlers) should be emitting MCEs; if you want to ingest with Kafka (as a side note you can also just avoid this can call the GMS directly via restli, though that is not recommended). As part of the data ingestion process the `gms` will emit an MAE if the datastore is updated. Think of it this way: MCE is an event that says "my source metadata changed; please update Datahub/gms". MAE says "the data in gms has been updated". You don't want to emit MAEs yourself because that is gms' job; only it should do it to help stay in sync.

Also check out https://github.com/linkedin/datahub/blob/master/docs/what/mxe.md

**Comment by mars-lan** [comment]
Assuming the questions has been answered. Please feel free to reopen if it's still unclear.
